07-13-16-language-plugins.md

```diff
@@ -0,0 +1,94 @@
+
+The process for deploying a code bundle to Tessel can be generalized like so:
+
+1. Determine Project Location in FS ->
+2. Run any compilation/pre-processing steps ->
+3. Bundle executable code into tarball ->
+4. Run any pre-execution steps ->
+5. Execute the code/binary ->
+6. Run any post-execution steps ->
+
+As such, there are six functions that need to be exported for any language deployment mechanism:
+
+`checkConfiguration`, `preBundle`, `tarBundle`, `preRun`, `shell`/`binary`, and `postRun`.
+
+### checkConfiguration
+```.js
+// checkConfiguration to confirm project deployment files
+exportables.checkConfiguration = (pushdir, basename, program) => {
+  // Should return the basename and program name of project
+  return {
+    basename,
+    program
+  };
+};
+```
+
+`checkConfiguration` should be able to accept the directory that was deployed with `t2 run ...` and return the path to the entrypoint of the project. With Rust, we do this by parsing the `Cargo.toml` and with JavaScript we parse the project's `package.json`. It is expected to return an object with a `basename` and optional `program` property (for the name of the project).
+
+### preBundle
+```.js
+// Optional function to process options prior to bundling project
+exportables.preBundle = function(options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+`preBundle` is an optional step for any processes that need to be run prior to the next step where you actually package the directory up to be sent to the Tessel 2. It's less critical for Rust where we just set some options parameters but in JavaScript is integral to the binary compilation stage (where we inject our pre-compiled binaries in place of existing project binary dependencies).
+
+### tarBundle
+```.js
+// This must implement a Promise that resolves with a Buffer
+// that represents a DIRECTORY containing the desired bundle.
+exportables.tarBundle = function(options) {
+  // Should return a Promise which eventually gets resolved with the bundle
+  return Promise.resolve(bundle);
+};
+```
+`tarBundle` is a required step which should resolve a `Promise` with a Buffer that represents the tarballed code or binary. The CLI will automatically untar the bundle when after it is deployed on the Tessel 2.
+
+In the case of Rust, this was the biggest addition. In order to compile Rust code for Tessel, we need to set up a special cross compiler that can compile a binary that runs on Tessel's MIPS architecture. The easiest solution for now is to have the CLI send out the project to a preconfigured remote [cross compilation server](https://github.com/tessel/rust-compilation-server). The server then compiles and returns the binary to the CLI.
+
+### preRun
+```.js
+// Optionally make changes on Tessel before starting the program
```

Comment:
"on Tessel" comes off as a bit vague to me, but I don't know of a better way to say it. "as root on Tessel" maybe?